### ISSUE TITLE
child_process cannot find module "app"

### ISSUE THREAD

**Issue by arthurakay** [atom/electron#1777]
I have an application in which I host an `express` server on a host defined by the user. If the user decides to change the port number (or other settings), I'd like to restart everything.

Originally I thought to use the `forever` module to handle this, but that gave me some problems. Instead I'm trying to simply use a child process as such:

```
var child = require('child_process'),
    path  = require('path');

function startChild() {
    var app = child.fork(
        path.join(__dirname, '_launch.js')
    );

    app.on('exit', function (code) {
        // when changing the settings, we restart the server/app
        if (code !== 99) {
            // but we still want to allow the user to kill the app manually
            process.exit();
        }

        startChild();
        app = null;
    });
}

startChild();
```

Needless to say `Electron` would execute this code first -- but inside `_launch.js` I would begin using various Electron modules (namely `app` and `menu`). However this approach immediately throws the error `Cannot find module 'app'` when it gets into `_launch.js`.

I've seen a few issues reported in this repo related to opening `Electron` within a child process... but nothing quite like what I'm attempting to do.

Any ideas on how to resolve this?

**Comment by paulcbetts** [comment]
It's probably better to just restart the app. But to be honest, I'd work to just get rid of the entire concept of hosting an embedded webserver, you shouldn't need to do this (though I realize that it sometimes makes porting a web app easier).

**Comment by arthurakay** [comment]
I'm not porting a web app - the server I create (inside Electron) utilizes Socket.io to connect apps in a traditional web browser with my Electron interface. In other words, my Electron app is a remote debugging tool connected to other client apps.

Using `forever` allowed me to restart the app after setting changes automatically - which worked fine locally, but after running a custom build of Electron with my app I saw all sorts of funky problems (namely an infinite loop creating child processes).

As I understand things, creating a vanilla "child process" works fine -- it just doesn't inherit some of the parent process' environmental naunces (namely these Electron-specific node modules).

I've tried all of the following variations as well, to no avail:

```
    var app = child.fork(
        '_launch.js',
        process.execArgv,
        {
            cwd : __dirname,
            env : process.execPath
            //env : process.env
            //env : { "ATOM_SHELL_INTERNAL_RUN_AS_NODE" : "0" }
        }
    );
```

All give me the same problems, though using `process.execPath` throws a different error:
  - App threw an error when running [TypeError: Cannot assign to read only property 'ATOM_SHELL_INTERNAL_RUN_AS_NODE'

Asking the user to manually restart the app isn't the end of the world -- but I still don't see why this shouldn't be possible.
